### DISCUSSION TITLE
Feature request: E-Mail Adresse

### DISCUSSION THREAD

**Discussion by JustFxDev** [creecros/Mailmagik#25] [Ideas]
Hi Creecros,

sorry, it's my again.

I would like to suggest to transfer the adressing to a Task, Comment or Project from the E-Mail adress to the subject of an E-Mail. Reason:

If i send an e-mail to Task #20 i use

`Task#20 <kb@mydomain.de>`

Works fine. Perfect.

If i send then an e-mail to `Task#21 <kb@mydomain.de>` my E-Mail Client (Thunderbird and i think Outlook & Co. will do this as well) gives my

`Task#20 <kb@mydomain.de>`

as the target adress. This is not a simple Text. It is an e-mail field which can not simply be changed. I have then to use the local menu and choose edit E-Mail (field)
![grafik](https://user-images.githubusercontent.com/82320201/222260729-d2f32d70-89dc-45ef-bb8b-e20ba870b588.png)

and have to change this. Each time i use an other task nr. This is because the client maintains a history of used e-mail adresses.

Furthermore it seams that `Task` is case sensitiv on the PlugIn side. I would suggest to have this not to be case sensitive `CommentOnTask` and `Project` as well..

This PlugIn is a real time saver for working with Kanboard. Great work! Thanks!

Cheers, Fx

**Comment by creecros** [comment]
I just tested with Thunderbird and Outlook, and have no idea what you are talking about. Works fine for me.

![image](https://user-images.githubusercontent.com/26339368/222270148-b4c8a24b-e8de-473b-b4c1-25b420bbf4cb.png)

ya, playing with this even more, i can't figure out your issue. everything works fine for me.

**Comment by creecros** [comment]
ya this is a horrible idea, unless @alfredbuehler disagrees. plays a large role in what emails its looking at as well, so if it has to now grab ALL emails, I dont see this as anything but negative.

**Comment by alfredbuehler** [comment]
This is a giant Annoyance from TB. 
I think, to avoid this, you have to disable the feature 
![image](https://user-images.githubusercontent.com/2079289/222355065-b17956b6-ba48-483f-8740-3d2156e6771d.png)
and then remove all Mailmagik addresses from the address books.

Or you can create and use templates for Mailmagik mails.

For dropping the case sensitivity: **strictly no**.

**Comment by JustFxDev** [comment]
> For dropping the case sensitivity: strictly no.

Ok, do not beat me....  😄  😄  😄 

Disabling collection of outgoing E-Mail adresses in Thunderbird is not an option. For me it would be ok, but the collegues...

The concept `Task#26 <email@mydomain.de>` has yet another disadvantage: I have to write the complete E-Mail adress each time i send something to Kanboard.  Anoying. Typos occur often. Outcomplete distrubs. Even more worse with long domain names.

Thanks for all your work on that one!

Cheers, Fx

**Comment by wallesch** [comment]
Hello everyone,

I also have problems with addressing via the recipient address in the mail system used here (Groupwise).

In the case of system-internal e-mails, the system replaces the recipient data with the system-internal data, which cannot be changed either.

Only (external) e-mails received via smtp would allow the use of the plugin.

My suggestion: Enclose all "tags" for controlling the plugin at the beginning of the subject within square brackets.

Example:
Subject: [Task#11 s:2023-03-02 d:2023-03-31 p:3]The actual subject

Would make the addressing independent of the e-mail system used and its settings.

cheers Walle

**Comment by creecros** [comment]
this is what I do actually, I am a typer. I type the email out every time I use it, so maybe that's why I don't understand. in fact, I'm the opposite, autocomplete annoys me.

if you parse from the subject, you will increase the time to scan the emails, considerably. it would also be a considerable rewrite of code. it would also change the behavior adversely for other people who would not want it this way. 

I could provide links that can be copy/pasted for each action, if that would help. maybe even `mailto:` links, would help.

I will think upon this more.

**Reply by JustFxDev** [reply]
> this is what I do actually, I am a typer. I type the email out every time I use it, so maybe that's why I don't understand. in fact, I'm the opposite, autocomplete annoys me.

Me, too 😄, but it feels not handy and smooth, nor right. Sorry.

And as @wallesch states it is a huge problem with some e-mail clients.

If it was only on me i would say:  i'am fine. I can handle this. But if you have a "normal" user (whatever that could be) in mind it is somehow "unfancy"(?) or generates problems that could not easy be fixed.

On the other hand i understand your concerns and you objections. Unsure how to go on.

My suggestion would be: let it depend on the topic of an e-mail (as @wallesch described). I use that only with actions that work in the background (thats my plan, have to setup that first). So i would not use it manually with the menu entry in the task details. Everything would happen in the background. I would not care about time to parse the topic. And i think the cron job for that should run at least once a hour. min. And in case of a manuell run: the spinner shows it's running...

But to be honest: i can not really judge this. 

> I could provide links that can be copy/pasted for each action, if that would help. maybe even `mailto:` links, would help.

I don't think that that makes sense. I would not do that. 
 
> I will think upon this more.

Always a good idea 😄 

Thanky you very much for your efforts on this one!

Cheers, Fx

**Reply by creecros** [reply]
I can definitely see both sides of the argument. But I wouldn't want to necessarily "change" it entirely. Giving the option to do it either way, could be a solution, and maybe one i'm willing to explore, but when that would be, i can't say.

**Reply by creecros** [reply]
so the good news is, I believe this is easier than I thought it would be. I think I can make a simple switch with 2 options, 1.) parse via `to:` 2.) parse via `subject:` and then it should be all set.

**Reply by alfredbuehler** [reply]
> so the good news is, I believe this is easier than I thought it would be. I think I can make a simple switch with 2 options, 1.) parse via `to:` 2.) parse via `subject:` and then it should be all set.

Great! I had exactly the same idea.

**Reply by creecros** [reply]
@alfredbuehler 
I did not finish, and got hung up when it got to the original parsing code, around line 156 in the action. For some reason it is parsing tags i think out of the subject....I pushed it where it is at, if you get time, otherwise, I will have to continue at another time. 

ea798dcca4d32365634bad97ca33dcd866e217be

here was my error:
```
Fatal error: Uncaught TypeError: Kanboard\Model\TaskModificationModel::updateTags(): Argument #2 ($original_task) must be of type array, null given, called in /var/www/app/app/Model/TaskModificationModel.php on line 27 and defined in /var/www/app/app/Model/TaskModificationModel.php:120
```

which made no sense to me, since `t:` was not in my subject line.
my subject line was `Project#1 Test what else I can say.`

**Reply by creecros** [reply]
@alfredbuehler nm i see the issue, it was staring me in the face and i just kept reading right over it

**Reply by alfredbuehler** [reply]
Argh, we did some parallel work. I thought you stopped.

**Reply by creecros** [reply]
i did, but my natural instinct to finish the job was too strong! i'm sorry!

**Comment by creecros** [comment]
@JustFxDev @alfredbuehler @wallesch 

Unless I missed something, which is possible it should work now. Give it a try.

![image](https://user-images.githubusercontent.com/26339368/222550776-a59370b8-af0f-46e6-a7cd-7dd2342ac19f.png)

**Reply by JustFxDev** [reply]
You guys are awesome!

I do not come at all behind with testing (Mailmagik, iCal, Metamagik <-> Task2Pdf) 😄 😄 

Unbelievable how much is happening here right now. Great!

Please state always what to test (latest Release or Code from the repo) - I will use the code from the repo for now.

I will test that and give feedback asap!

Cheers, Fx

**Comment by JustFxDev** [comment]
You asked for it 😄 😄 😄 

Mode: _Use subject_
No cronjobs or actions defined so far
PlugIn Version: V1.2.2
Kanboard: 1.2.25


### 0.) Fatal Error

An incorrect imap e-mal password or something like that disables the plug in. It disappears from the list of installed plugins. This leads to this situation: I cannot correct the password in the application setting / e-mal setup. The fields are no longer there. 

Even a new installation does not change this, because the wrong information is still in the database. 

I fixed that by changing it directly in the database.

That occurred also as the server had certification problems (this was the reason why I found that initial out)

Error-log:
[xx-xxx-xxxxx xx:xx:xx Europe/Berlin] [critical] Mailmagik: ["Can not authenticate to IMAP server: [AUTHENTICATIONFAILED] Authentication failed."]

### 1.) Processing an e-mail manually via task menu entry
Send an E-Mail with Subject „Task#84” to Kanboard. When I manually start processing I get this page in this browser. No buttons, no way to get out beside “go back” browser function works (two times).

![grafik](https://user-images.githubusercontent.com/82320201/222667377-24795b2d-52c0-4703-ac17-dae38e308710.png)

The text says: “This information could not be found in the database”

That did not happen in the last version (that one without topic support). But even with this message the comment is applied to the task.

### 2.) New Task (understanding)
If I use a Task ID which is not in the database I get a new Task (expected, ok) with the content of the e-mail as description and the topic of the e-mail as Header of the task (expected, ok).

But I receive also the same information as comment in the new task.

I would suggest: when creating a new task, no comment is created.

### 3.) Hover on buttons
Suggestion: Display a hint what the button does, when hover over the buttons would be nice
![grafik](https://user-images.githubusercontent.com/82320201/222667513-2ab81296-d344-46d0-bf8b-bb955e9781a9.png)

### 4.) When I delete an e-mail and confirm “yes” the dialog does not close (infinite):

![grafik](https://user-images.githubusercontent.com/82320201/222667700-73a1fddb-b40d-47ff-ac7b-53b6b367ca91.png)

When this occurs I select “Cancel” (Abbrechen) und reload the task. The e-mail is gone. Seams the mail will be deleted, but after that something causes to hang.


### 5.) e-mail to project
When I send an e-mail with subject “Project#3” (an existing one): where does that mail go to? Can’t find it.

![grafik](https://user-images.githubusercontent.com/82320201/222669599-592387cb-406f-455a-8048-46bbd3690368.png)


### 6.) Matter of understanding:

a.)
I send an e-mail with topic “Task#86” – Existing one. I get a new comment

b.)
I send an e-mail with topic “CommentOnTask#86” – Existing one. I get a new comment exactly like a.)

What’s the difference?


### 7.) Feature / Future Ideas for further development 😄 

a.)
Send an e-mail to create a task in a specific column. Maybe in such a manner:

In Board/Settings/Columns I may see beside the name of the column an automatic created “target e-mail tag” like this (the number 3 is the project-id):

![grafik](https://user-images.githubusercontent.com/82320201/222667784-6821363e-3832-497b-ae7d-b38ba1866dc9.png)

For the topic I use then

NewTaskInColumn#backlog-3 My  very important task topic

### 8.) Conclusion (maybe! Only as a suggestion)

One may only(?) need

- Create new task in a project
-- Send task to a column in a project see 7.
-- NewTaskInColumn#ColumnID

- Create a comment to a task
-- CommentOnTask

- Change an existing Task
-- duedate etc. 

Everything combined with automatic actions and attachment handling (already available)

### 9.) e-mail sender restrictions
Restrict e-mails to a board to known e-mails addresses from users of that board (should be configurable yes/no would be sufficent)

**Comment by creecros** [comment]
`#0` fixed

for `#1` and `#2` I do not know what you mean, about Comments being created. I do not get this behavior. 

`#5` Assuming there is a Project#3 and You have an Action enabled in Project#3 to convert an email to a task, it creates a task in project#3.

`#6a` Sending an email with Subject `Task#86` should do nothing, automatically. I will look into it, but that would be how you send to the Task Email function, and should only show up there.

`#4` I see, there is something wrong and have not quite figured out why. This behavior exists no matter which parse method you use.

`#7` should be no problem adding to the parsing that currently exists
|in Subject|Effect
|-|-
|d:YYYY-MM-DD| Set due date
|s:YYYY-MM-DD |Set start date
|p:n |Set priority n
|c:name| Set category 'name'
|t:name| Add tag 'name', may be repeated. Non-existing tags will be created, beware of typos.

Just add the option for column

`#9` already exists but there is no option, emails must only come from users in the board.

**Comment by creecros** [comment]
Actually `#1` and `#2` and `#6a` although, your experience does not make sense, i do see an issue with the logic. if you use `CommentOnTask#` then the logic for `Task#` will pick that up. Does not explain why `Task#` would create a comment though.

**Comment by creecros** [comment]
0 - Fixed
1,2,6a - Fixed by adding bracket requirement for action. example subject, `[Project#1] My subject d:2023-02-27 c:SQL t:sql` 
4 - Fixed

Everything else should be additions, but I think the bugs are now worked out.

**Reply by alfredbuehler** [reply]
But I would wait with a release. Some code for the subject scanning needs more care, IMHO

**Reply by creecros** [reply]
@alfredbuehler Go ahead, i am done for the day.

**Reply by creecros** [reply]
I will let you add this "care", you will be better at that than myself. I'll hold on release until you give the ok.

**Reply by alfredbuehler** [reply]
I did some inspection and additional testing. IMHO it's all OK, I have nothing to amend for now. You can go on.

**Comment by pahles** [comment]
I've tested this and everything seems to be working! Keep up the good work.
